drivers/clock_control/clock_control_nrf2_global_hsfll.c

```diff
@@ -0,0 +1,302 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT nordic_nrf_hsfll_global
+
+#include "clock_control_nrf2_common.h"
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/clock_control/nrf_clock_control.h>
+#include <nrfs_gdfs.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_DECLARE(clock_control_nrf2, CONFIG_CLOCK_CONTROL_LOG_LEVEL);
+
+#define GLOBAL_HSFLL_CLOCK_FREQUENCIES \
+	DT_INST_PROP(0, supported_clock_frequencies)
+
+#define GLOBAL_HSFLL_CLOCK_FREQUENCIES_IDX(idx) \
+	DT_INST_PROP_BY_IDX(0, supported_clock_frequencies, idx)
+
+#define GLOBAL_HSFLL_CLOCK_FREQUENCIES_SIZE \
+	DT_INST_PROP_LEN(0, supported_clock_frequencies)
+
+#define GLOBAL_HSFLL_FREQ_REQ_TIMEOUT \
+	K_MSEC(CONFIG_CLOCK_CONTROL_NRF2_GLOBAL_HSFLL_TIMEOUT_MS)
+
+#define GLOBAL_HSFLL_INIT_LOW_REQ \
+	CONFIG_CLOCK_CONTROL_NRF2_GLOBAL_HSFLL_REQ_LOW_FREQ
+
+BUILD_ASSERT(GLOBAL_HSFLL_CLOCK_FREQUENCIES_SIZE == 4);
+BUILD_ASSERT(GLOBAL_HSFLL_CLOCK_FREQUENCIES_IDX(0) == 64000000);
+BUILD_ASSERT(GLOBAL_HSFLL_CLOCK_FREQUENCIES_IDX(1) == 128000000);
+BUILD_ASSERT(GLOBAL_HSFLL_CLOCK_FREQUENCIES_IDX(2) == 256000000);
+BUILD_ASSERT(GLOBAL_HSFLL_CLOCK_FREQUENCIES_IDX(3) == 320000000);
+BUILD_ASSERT(GDFS_FREQ_COUNT == 4);
+BUILD_ASSERT(GDFS_FREQ_HIGH == 0);
+BUILD_ASSERT(GDFS_FREQ_MEDHIGH == 1);
+BUILD_ASSERT(GDFS_FREQ_MEDLOW == 2);
+BUILD_ASSERT(GDFS_FREQ_LOW == 3);
+
+struct global_hsfll_dev_config {
+	uint32_t clock_frequencies[GLOBAL_HSFLL_CLOCK_FREQUENCIES_SIZE];
+};
+
+struct global_hsfll_dev_data {
+	STRUCT_CLOCK_CONFIG(global_hsfll, GLOBAL_HSFLL_CLOCK_FREQUENCIES_SIZE) clk_cfg;
+	const struct device *dev;
+	struct k_work evt_work;
+	nrfs_gdfs_evt_type_t evt;
+	struct k_work_delayable timeout_dwork;
+
+#if GLOBAL_HSFLL_INIT_LOW_REQ
+	struct k_sem evt_sem;
+#endif /* GLOBAL_HSFLL_INIT_LOW_REQ */
+};
+
+static uint32_t global_hsfll_get_max_clock_frequency(const struct device *dev)
+{
+	const struct global_hsfll_dev_config *dev_config = dev->config;
+
+	return dev_config->clock_frequencies[ARRAY_SIZE(dev_config->clock_frequencies) - 1];
```

Comment:
This could be returned build-time since there are build asserts checking for supported clock frequencies?

Reply:
this driver does have some strangeness regarding single instances / multi instance code, I wanted to make it as easy as possible to extend it if we choose to support more HSFLLs in the future